### ISSUE TITLE
[Slim 4] remove App::setContainer as it introduces a bug where the CallableResolver does not know about the container

### ISSUE THREAD

**Issue by lordrhodos** [slimphp/Slim#2566]
The curren App constructor supports passing an optional PSR Container:

https://github.com/slimphp/Slim/blob/bcf7df5129a3d126999cc73a6c961d35b83fce35/Slim/App.php#L87-L95

The `App::getCallableResolver` uses the container on the first call to initiate the callableResolver
https://github.com/slimphp/Slim/blob/bcf7df5129a3d126999cc73a6c961d35b83fce35/Slim/App.php#L210-L217

The `App::setContainer` allows to set the container outside the constructor but does not set the container at the existing callable resolver. This can lead to a situtation where the CallableResolver does not know about the container added via `$app->setContainer($container)`.

Of course the container could be set on the callable resolver when set via the `setContainer`method in App but I suggest to remove the container setter completly. As the container should be passed (optionally) via the Apps constructor I do not see any reason why a setter should be provided.

**Comment by lordrhodos** [comment]
just saw that the middlware PSR-15 PR https://github.com/slimphp/Slim/pull/2555

is dealing with this (https://github.com/slimphp/Slim/pull/2555/files#diff-eb6c3386171c6f5318119100a52f3db7R123) which I think creates another issue. Imagine something like:
 
```php
$container = new Container();
$callableResolver = new CustomCallableResolver();
$app = new Slim\App($responseFactory, null, $settings);
$app->setCallableResolver($callableResolver);
$app->setContainer($container);
```

In this case the `CustomCallableResolver` would be overriden by Slims default CallableResolver or am I wrong?

**Comment by l0gicgate** [comment]
I just fixed this logic error on #2555 with [this commit](https://github.com/slimphp/Slim/pull/2555/commits/6a95a4c3caa30ebe687dc72a439480eb10468a46). I also added a `setContainer()` method on `CallableResolverInterface` so you can set the container on the fly which `App::setContainer()` is now using.

Good catch! Let me know if you want to close this or if you want me to.

**Comment by lordrhodos** [comment]
👍 nice! Hope you get #2555 merged soon, seems like it is growing and growing 😄 

I'll close it.
